Fix multiple operator rule in PHP lexer

pyrmont opened the pull request:
PHP has a rule that can match one or more symbols as operators. Although this permits combinations that are not syntactically correct, it provides some performance improvement for code that has sequential operators.

The problem is that two of these operators, `?>`, have a special meaning in PHP. Matching multiple operators in one pass causes the `?>` to be lexed incorrectly if it is immediately preceded by an operator. This is possible with the `:` operator. This PR fixes the rule by splitting it between operators that can appear multiple times in sequence and those that cannot.

This fixes #1362.

pyrmont commented:
@HypeMC, @julp: How does this look to you?

julp commented:
Is it a problem if operators like `??` or `??=` are highlighted as several tokens?

pyrmont commented:
@julp Not really. The issue isn't about the number of tokens that are output; Rouge will merge the identical contiguous tokens together in the final output. The issue is the number of characters that the regular expression matches. The earlier rule was matching one or more operator symbols but since these symbols included `:`, `?` and `>`, it meant that `:?>` would get matched by this rule and `?>` would never get matched by the rule for closing PHP tags.

julp commented:
In fact, only `?` is problematic, right? We could achieve the same with just `rule %r/[?]/`, couldn't we? Whatever you can find before `?>` the single fact that `?` is apart garantee you to not have a match longer than `?>` and be consumed instead of it.

pyrmont commented:
@julp Good point about the `?` being the real issue. I've rebased on the just-committed PHP lexer and split that out into its own rule. What do you think?

julp commented:
Yes unless you want to use a negative assertion? (eg `%r/([~!%^&*+=\|:.<>\/@-]|(\?(?!>)))+/`) Which, IMO, would be more error prone and less readable.

pyrmont commented:
@julp Agreed. Will merge this in now.

pyrmont commented:
@julp Thanks for your help! Another bug fixed! :)
